### PR TITLE
macOS

### PR DESCRIPTION
**Pull request by mhuusko5**
Per #21, WIP macOS support for the frameworks and example/test suite.

So far Core and Database have been setup, and are building/tests passing (Database_IntegrationTests are not, but they weren't before these commits.. 🤷‍♂️).

Auth and Storage are to come. It's just a matter of isolating/pulling macOS compatibility changes from https://github.com/firebase/firebase-ios-sdk/pull/22, as well as some annoying/fiddly project/config work, so before I get too far with that I want to make sure I'm not doing anything crazy.

Mmm, also: I'm not quite sure what it would take to get AllUnitTests_macOS going on Travis along with AllUnitTests_iOS. EDIT: Think I figured it out. Not sure what purpose https://github.com/firebase/firebase-ios-sdk/blob/master/.travis.yml#L8 serves though since test.sh seems to do the work.

### CODE REVIEW

**Comment by paulb777** [comment]
Thanks @mhuusko5

The IntegrationTests are not automated, but should run with the instructions at https://github.com/firebase/firebase-ios-sdk#running-sample-apps.

Yeah, we should probably delete https://github.com/firebase/firebase-ios-sdk/blob/master/.travis.yml#L8 - @ulukaya ?

I'll review code changes later.

**Comment by mhuusko5** [comment]
Btw, the macOS tests are failing to start on Travis with error 65 (the test.sh retry nor an additional sleep I added seem to help), `Early unexpected exit, operation never finished bootstrapping - no restart will be attempted`. I know it's a common error, but I'm not sure why it would be happening outside of simulator usage. Maybe someone on your end has experience.

**Comment by mhuusko5** [comment]
@paulb777 mmm, quite odd. I'm not sure. Shall investigate.
So running the example/tests project and test.sh fail for you also, or are you just seeing this while linking manually like in your gist?

**Comment by paulb777** [comment]
The gist is a cut and paste from Xcode. I see the error both for iOS and OSX with all targets - since it's a Core dependency.  

Since it runs for you and on travis, it might be some configuration thing for me, but I don't know what.

**Comment by mhuusko5** [comment]
Ah, right right.
Well, I haven't run into this exact issue, but I was experiencing some trouble for a while earlier with the GTMSessionFetcher dependency and some seriously mysterious errors. Wiping DerivedData + clang/Xcode caches under /var/folders is what finally did it for me.
If that doesn't work, I'd be curious to know if a pod deintegrate; pod install makes any difference, although it's not an optimal solution given that it would likely result in changes to commit.
Lastly... a fresh clone, if this wasn't already.

Sorry I only have the standard advice!

Oh, last thing that could be different/above wouldn't cover. What do you have under ~/.cocoapods/config.yaml?

**Comment by paulb777** [comment]
Nothing worked until I did a fresh clone. Once I did that, I realized that pod update/install was not previously coloring FirebaseDev green, even after "pod deintegrate" and removing the Pods directory

I just successfully ran "AllUnitTests_macOS"!!!!

**Comment by paulb777** [comment]
Making the test scheme runnable and splitting the build from test might be a travis solution - https://gist.github.com/mattyohe/31552b6543ae0cf2fdca912c29e9bb17#file-makefile

File: FirebaseDev.podspec [hunk 1]
```diff
@@ -16,6 +16,7 @@ Simplify your iOS development, grow your user base, and monetize more effectivel
   s.source           = { :git => 'https://github.com/firebase/firebase-ios-sdk.git', :tag => s.version.to_s }
   s.social_media_url = 'https://twitter.com/Firebase'
   s.ios.deployment_target = '8.0'
+  s.osx.deployment_target = '10.10'
```
Comment (paulb777) [inline FirebaseDev.podspec]: Since we're not yet building and publishing the iOS CocoaPod from the open source repo, it might also be useful to make a separate MacOS podspec in the interim.

Comment (mhuusko5) [inline FirebaseDev.podspec]: If you mean in order to publish early, then instead of duplicating this, disabling iOS, and publishing as FirebaseMac, how about just pushing this one as-is as FirebaseDev? Then people who are eager can test both sides, integrate in cross platform frameworks, etc.

**Comment by mhuusko5** [comment]
FirebaseAuth is all set, so this PR is probably all set afaict, except for the Travis issue: still no luck getting the macOS tests to run/not 65, even with pre-test build, retry, delay. I find it suspicious too that it's the macOS ones that won't work, when they don't even need a warm up period/there's no simulator involved. Makes me think maybe this error 65 situation isn't the usual one.

**Comment by mhuusko5** [comment]
For the time being, I've disabled the iOS tests as well as xcpretty so Travis spits out some more/helpful info. The below is happening for each framework and is probably the culprit of the early exit/65. Not sure of a solution yet (or why OCMock wouldn't be loading properly in the first place; just for macOS).

```
2017-05-28 12:03:51.150798+0000 Core_Example_macOS[8227:18353] Failed to load test bundle from file:///Users/travis/Library/Developer/Xcode/DerivedData/Firebase-bwsvaxcjgoxobgeyrakpixqtsyty/Build/Products/Debug/Core_Example_macOS.app/Contents/PlugIns/Core_Tests_macOS.xctest/: Error Domain=NSCocoaErrorDomain Code=3587 "dlopen_preflight(/Users/travis/Library/Developer/Xcode/DerivedData/Firebase-bwsvaxcjgoxobgeyrakpixqtsyty/Build/Products/Debug/Core_Example_macOS.app/Contents/PlugIns/Core_Tests_macOS.xctest/Contents/MacOS/Core_Tests_macOS): Library not loaded: @rpath/OCMock.framework/Versions/A/OCMock

  Referenced from: /Users/travis/Library/Developer/Xcode/DerivedData/Firebase-bwsvaxcjgoxobgeyrakpixqtsyty/Build/Products/Debug/Core_Example_macOS.app/Contents/PlugIns/Core_Tests_macOS.xctest/Contents/MacOS/Core_Tests_macOS

  Reason: image not found" UserInfo={NSLocalizedFailureReason=The bundle is damaged or missing necessary resources., NSLocalizedRecoverySuggestion=Try reinstalling the bundle., NSFilePath=/Users/travis/Library/Developer/Xcode/DerivedData/Firebase-bwsvaxcjgoxobgeyrakpixqtsyty/Build/Products/Debug/Core_Example_macOS.app/Contents/PlugIns/Core_Tests_macOS.xctest/Contents/MacOS/Core_Tests_macOS, NSDebugDescription=dlopen_preflight(/Users/travis/Library/Developer/Xcode/DerivedData/Firebase-bwsvaxcjgoxobgeyrakpixqtsyty/Build/Products/Debug/Core_Example_macOS.app/Contents/PlugIns/Core_Tests_macOS.xctest/Contents/MacOS/Core_Tests_macOS): Library not loaded: @rpath/OCMock.framework/Versions/A/OCMock

  Referenced from: /Users/travis/Library/Developer/Xcode/DerivedData/Firebase-bwsvaxcjgoxobgeyrakpixqtsyty/Build/Products/Debug/Core_Example_macOS.app/Contents/PlugIns/Core_Tests_macOS.xctest/Contents/MacOS/Core_Tests_macOS

  Reason: image not found, NSBundlePath=/Users/travis/Library/Developer/Xcode/DerivedData/Firebase-bwsvaxcjgoxobgeyrakpixqtsyty/Build/Products/Debug/Core_Example_macOS.app/Contents/PlugIns/Core_Tests_macOS.xctest, NSLocalizedDescription=The bundle “Core_Tests_macOS.xctest” couldn’t be loaded because it is damaged or missing necessary resources.}
```

File: FirebaseDev.podspec (hunk 1 above)
Comment (paulb777) [inline FirebaseDev.podspec]: Good point. The downside is that the FirebaseDev pod is incompatible with the Firebase pod and could cause confusion, especially for developers incrementally wanting to add iOS Firebase features not currently open sourced. I'll talk to the product team next week.

**Comment by paulb777** [comment]
This is the closest potential solution to the OCMock travis problem I can see - https://stackoverflow.com/questions/14760435/adding-ocmock-causes-test-to-launch-main-app-instead-of-running-tests

File: FirebaseDev.podspec (hunk 1 above)
Comment (mhuusko5) [inline FirebaseDev.podspec]: Mm, right right. On that note, curious if it might make sense to fill in FirebaseDev with subspecs/dependencies for the missing static components (RemoteConfig, AdMob, etc.)? Then FirebaseDev could actually be used in place of Firebase, existing on trunk distinctly until the final call on static/dynamic build/distribution is made (and until more of the subspecs are open sourced, I imagine).

Comment (paulb777) [inline FirebaseDev.podspec]: Unfortunately, that wouldn't work because of duplication between the FirebaseDev version of FirebaseCore and the published static library FirebaseCore that is a dependency of the missing static components.

Comment (mhuusko5) [inline FirebaseDev.podspec]: Ah, spaced that, silly me.

**Comment by mhuusko5** [comment]
Success!

**Comment by paulb777** [comment]
Nice! Great work!

**Comment by paulb777** [comment]
I'm able to run the Database integration tests, but something is wrong with the configuration of the Storage integration tests: 
screen shot 2017-05-28 at 4 25 30 pm

**Comment by mhuusko5** [comment]
I ran into that for various targets a few times last weekend. I believe restarting Xcode/clearing DerivedData did the trick; perhaps more fiddling, but pretty sure no actual changes necessary.

**Comment by ulukaya** [comment]
We can delete line 7 and 8 from .travis.yml

**Comment by mhuusko5** [comment]
@paulb777 looks like there *was* a config problem (wrong sdk) with the Storage integration tests, preventing building. Hopefully it will resolve the TEST_HOST issue too.

**Comment by paulb777** [comment]
Yep. That fixes it for me. Thanks @mhuusko5 !

I'm now able to run both the Storage and Database Integration tests.  I see one failure with the Database tests in testSettingANodeWithChildrenToAPrimitiveAndBack, but it happens for both iOS and macOS and it goes away with a breakpoint, so I suspect it's a flaky test issue with too small of kFirebaseTestWaitUntilTimeout.

The only remaining implementation issue I see is, what if anything to do about the [AuthSamples](https://github.com/firebase/firebase-ios-sdk/tree/master/AuthSamples), which are UIKit based.

**Comment by mhuusko5** [comment]
Indeed, AuthSamples is non-trivially tied to UIKit/iOS, and separating out the platform agnostic logic then creating a macOS sample (including some UI work, it seems) will take a little time. I could start to take a look at it later in the week, but I think it might best be a separate PR.

**Comment by paulb777** [comment]
Agreed on a separate PR for AuthSamples.

Since this is a large PR, I'm asking for reviews and LGTM's from Auth, Storage, Database, Core and product management teams before merging.

**Comment by paulb777** [comment]
To try out this PR:

> ~  $ mkdir pr-test
~  $ cd pr-test/
~/pr-test  $  git clone https://github.com/mhuusko5/firebase-ios-sdk.git
~/pr-test  $ cd firebase-ios-sdk/
~/pr-test/firebase-ios-sdk (master) $ git checkout macOS
~/pr-test/firebase-ios-sdk (macOS) $ cd Example/
~/pr-test/firebase-ios-sdk/Example (macOS) $ pod update
~/pr-test/firebase-ios-sdk/Example (macOS) $ open Firebase.xcworkspace/

Then run and test the schemes.

File: Firebase/Auth/Source/FIRAuth.m [hunk 2]
```diff
@@ -23,18 +23,14 @@
 #import "AuthProviders/EmailPassword/FIREmailPasswordAuthCredential.h"
 #import "AuthProviders/Phone/FIRPhoneAuthCredential_Internal.h"
 #import "Private/FIRAdditionalUserInfo_Internal.h"
-#import "Private/FIRAuthAPNSToken.h"
-#import "Private/FIRAuthAPNSTokenManager.h"
 #import "Private/FIRAuthAppCredentialManager.h"
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/FIRAuth.m]: FIRAuthAppCredentialManager is only used for Phone number auth, so please move it to the inside of `#if TARGET_OS_IOS` as well.

File: Firebase/Auth/Source/FIRAuth.m [hunk 3]
```diff
@@ -376,14 +386,21 @@ - (instancetype)initWithAPIKey:(NSString *)APIKey appName:(NSString *)appName {
...
                   @"Error loading saved user when starting up: %@", error);
     }
+
+    #if TARGET_OS_IOS
     // Initialize for phone number auth.
     _tokenManager =
         [[FIRAuthAPNSTokenManager alloc] initWithApplication:[UIApplication sharedApplication]];
+    #endif
+
     _appCredentialManager = [[FIRAuthAppCredentialManager alloc] initWithKeychain:_keychain];
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/FIRAuth.m]: Please include all references to _appCredentialManager to the inside of `#if TARGET_OS_IOS`.

File: Firebase/Auth/Source/RPCs/FIRAuthBackend.h [hunk 4]
```diff
@@ -464,6 +466,7 @@ typedef void (^FIRVerifyClientResponseCallback)
 - (void)sendVerificationCode:(FIRSendVerificationCodeRequest *)request
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/RPCs/FIRAuthBackend.h]: Please include this method in `#if TARGET_OS_IOS` as it is used by Phone Number Auth only.

File: Firebase/Auth/Source/RPCs/FIRAuthBackend.m [hunk 5]
```diff
@@ -379,10 +382,12 @@ + (void)sendVerificationCode:(FIRSendVerificationCodeRequest *)request
...
 }
 
+#if TARGET_OS_IOS
 + (void)verifyPhoneNumber:(FIRVerifyPhoneNumberRequest *)request
                  callback:(FIRVerifyPhoneNumberResponseCallback)callback {
   [[self implementation] verifyPhoneNumber:request callback:callback];
 }
+#endif
 
 + (void)verifyClient:(id)request callback:(FIRVerifyClientResponseCallback)callback {
```
Comment (XiangtianDai) [inline Firebase/Auth/Source/RPCs/FIRAuthBackend.m]: Please include this method in `#if TARGET_OS_IOS`.

File: Firebase/Auth/FirebaseAuth.podspec [hunk 6]
```diff
@@ -20,8 +20,16 @@ Simplify your iOS development, grow your user base, and monetize more effectivel
   s.source           = { :git => 'https://github.com/firebase/firebase-ios-sdk.git', :tag => s.version.to_s }
   s.social_media_url = 'https://twitter.com/Firebase'
   s.ios.deployment_target = '7.0'
+  s.osx.deployment_target = '10.10'
 
   s.source_files = '**/*.[mh]'
+  s.osx.exclude_files =
```
Comment (XiangtianDai) [inline Firebase/Auth/FirebaseAuth.podspec]: And FIRAuthAppCredentialManager here, too, for it is only used by Phone Number Auth.

File: Example/Firebase.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -1486,7 +2054,7 @@
 				CLASSPREFIX = FIR;
 				LastSwiftUpdateCheck = 0830;
 				LastUpgradeCheck = 0820;
-				ORGANIZATIONNAME = "Paul Beusterien";
```
Comment (salqadri) [inline Example/Firebase.xcodeproj/project.pbxproj]: lol!

File: Firebase/Core/FIRAppEnvironmentUtil.m [hunk 8]
```diff
@@ -20,6 +20,8 @@
 #import <mach-o/dyld.h>
 #import <sys/utsname.h>
 
+NS_ASSUME_NONNULL_BEGIN
```
Comment (davidair) [inline Firebase/Core/FIRAppEnvironmentUtil.m]: Is this change required? If it's not needed for macOS support, I'd keep it out.

File: BuildFrameworks/build.swift [hunk 9]
```diff
@@ -110,17 +119,26 @@ func syncExec(command: String, args: [String] = []) {
...
-func buildThin(framework: String, arch: String, sdk: String, parentDir: String) -> [String] {
+func buildThin(framework: String, multiplatform: Bool, arch: String, multisdk: Bool, sdk: String, parentDir: String) -> [String] {
+  let schemeSuffix: String
+  if !multiplatform {
+    schemeSuffix = ""
+  } else if sdk.hasPrefix("mac") {
+    schemeSuffix = "-OSX"
+  } else if sdk.hasPrefix("iphone") {
+    schemeSuffix = "-iOS"
+  } else { fatalError("TODO: tvOS/watchOS") }
```
Comment (rsattar) [inline BuildFrameworks/build.swift]: Any reason this is not on a new line, like the rest of the if-else blocks?

File: Firebase/Database/Core/FRepo.m [hunk 10]
```diff
@@ -50,7 +50,7 @@
 #import "FValueEventRegistration.h"
 #import "FEmptyNode.h"
 
-#ifdef TARGET_OS_IPHONE
+#if TARGET_OS_IPHONE
```
Comment (rsattar) [inline Firebase/Database/Core/FRepo.m]: This was copied from before, but shouldn't this be TARGET_OS_IOS ?

**Comment by rsattar** [comment]
LGTM from Messaging, left some minor comments. I'm curious to see what it'll take to get Push Notification support in macOS!

File: BuildFrameworks/build.swift (hunk 9 above)
Comment (mhuusko5) [inline BuildFrameworks/build.swift]: Just a style, happy to change.

File: Firebase/Database/Core/FRepo.m (hunk 10 above)
Comment (mhuusko5) [inline Firebase/Database/Core/FRepo.m]: It could be, yes. There's ~12 other uses of TARGET_OS_IPHONE in other files. Shall I change those while I'm at it?

**Comment by mhuusko5** [comment]
@davidair @XiangtianDai btw, comments addressed. thanks!

**Comment by davidair** [comment]
LGTM for Core

**Comment by XiangtianDai** [comment]
LGTM for Auth

**Comment by schmidt-sebastian** [comment]
LGTM for Database.

The Storage integration tests for Mac only worked for me after I compiled the iOS tests first - before that the build failed due to missing files (FirebaseStorage.h for example). I suspect we are missing a target dependency somewhere.

I also had to remove the 'Nullable_' specifier in FIRLogger.h:87 .

**Comment by mhuusko5** [comment]
@schmidt-sebastian interesting.. on the FIRLogger issue, that is something I had fixed (needs to be `TARGET_OS_SIMULATOR || TARGET_OS_OSX`) in a previous incarnation of this PR, but didn't bring that change over because everything was suddenly building without it. Not sure what may be different now/for you, but I guess I'll just add it back.

On the Storage tests build issue (/weird temporary build issues in general), there may be some specific/isolated cause (will do a comb through; I'm not personally running into this one though), but it's also possible it falls under the umbrella of general CocoaPods/header finicky-ness. I've experienced header related build quirks with iOS/macOS targets/CP dependencies living side by side in projects before, but then things definitely got even more finicky over the weekend when I had to some odd manual/duplicative linking of OCMock to appease an issue that was only happening on Travis. The Travis issue is something I'd like to revisit in the future; I think it's going to need some help or info from them as two what might be special about their environments.
    In the meantime though/the good news, is for *me*, any clean build issue/failure I have experienced has been resolved by a second build attempt (could you try that for the Storage integration issue you're experiencing; I'd be curious if that does it or with your environment if you *have* to build iOS first).

**Comment by paulb777** [comment]
@schmidt-sebastian @mhuusko5 I'm also able to build the Storage Integration Tests for Mac from a clean clone.

FYI to all, the goal is to merge the PR tomorrow(Friday).  I'll verify that there are no regressions.  Then we can address any lingering issues in much smaller PRs.

**Comment by schmidt-sebastian** [comment]
I built the Storage tests again from a clean repo and am able to run them without any modifications or prior builds. I am ok if we fix this issue at a later point should it resurface.

LGTM for Storage & Database.

**Comment by mhuusko5** [comment]
🎊 🎊 🎊 🎊 🎊 

Thanks for the help, all!
